### PR TITLE
3.0: Make rpm packaging work again

### PR DESCRIPTION
**Pull request by DrDaveD**
**Description of the Pull Request (PR):**

This PR makes rpm packaging work again for the master branch

**This fixes or addresses the following GitHub issues:**

- Ref: #

**Checkoff for all PRs:**

- [x] I have read the [Guidelines for Contributing](https://github.com/singularityware/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- [ ] I have added changes to the [CHANGELOG](https://github.com/singularityware/singularity/blob/development/CHANGELOG.md) and and documentation updates to the [singularityware](https://www.github.com/singularityware/singularityware.github.io) documentation base.
- [x] I have tested this PR locally with a `make test`
- [x] This PR is against the project's master branch
- [x] I have added myself as a contributor to the [contributors's file](https://github.com/singularityware/singularity/blob/master/CONTRIBUTORS.md)
- [ ] This PR is ready for review and/or merge

Attn: @singularityware-admin

### CODE REVIEW

**Comment by DrDaveD** [comment]
To build an rpm with this PR, you run dist/rpm/rpmbuild any time after running mconfig, on an rpm-based Linux distro.  I tested on Scientific Linux 6 & 7, and plan to put centos 6 & 7 builds into travis after this is merged.

**Comment by DrDaveD** [comment]
Rebased on current master.

@yhcote could you please review?  Most of the changes are in mconfig (not counting singularity.spec.in which was non-functional).

File: mlocal/frags/Makefile.stub
Comment (jmstover) [inline mlocal/frags/Makefile.stub:19]: We're _wanting_ to get all the base autoconf options added in. So...  This _should_ end up being:
`$(BINDIR)/singularity`

Where BINDIR defaults to `$(EXEC_PREFIX)/bin` ... and `$(EXEC_PREFIX)` defaults to `$(PREFIX)`

In general, most of these options:
https://www.gnu.org/prep/standards/html_node/Directory-Variables.html

The idea is to pass to `mconfig` in the `%build` section of the spec like: 
`./mconfig -p %{_prefix} -e %{_exec_prefix} -l %{_libexecdir} -D %{_datadir}`
etc, etc... 

So we are using site specific RPM macro's when building.

When using autoconf, it was done automagically.

File: dist/rpm/singularity.spec.in
Comment (jmstover) [inline dist/rpm/singularity.spec.in]: In our distro released SPEC file, this was going to be a single RPM package. A very simple, basic, and _example_ SPEC file. 

These are decisions to be made downstream by the package maintainers, not by us. They should be in control of the SPEC file they use and not using ours.

Are you willing to change it so this generates a _single_ RPM package?

File: dist/rpm/singularity.spec.in
Comment (jmstover) [inline dist/rpm/singularity.spec.in]: Is this right? You're saying it `Provides` ${name}-runtime then `Obsoletes` ${name}-runtime?

I can see the `Obsoletes` here... but not the `Provides`. :/

**Review by jmstover — CHANGES_REQUESTED** [review]
Change RPM spec to be a _single_ package instead of multiple.

Other questions/comments

File: mlocal/frags/Makefile.stub
Comment (DrDaveD) [reply mlocal/frags/Makefile.stub:19]: Ok, but is this good enough as an intermediate step?

File: dist/rpm/singularity.spec.in
Comment (DrDaveD) [reply dist/rpm/singularity.spec.in]: I am the primary downstream package maintainer, for EPEL.  My preference is to make it be as close to the upstream as possible, so that people who compile from upstream source will get as similar an experience as possible.  So I would change it to a single package if I must, but I would rather not, because EPEL will have the dual package.

File: dist/rpm/singularity.spec.in
Comment (DrDaveD) [reply dist/rpm/singularity.spec.in]: Yes. that's right.  If someone currently has installed only singularity-runtime and they do a yum upgrade I want them to get the new singularity.  The purpose of singularity-runtime was to minimize exposure to setuid while still allowing a complete runtime environment, without necessarily including the ability to create images.  @bauerm97  told me that there's no privileged functionality used in 3.0 for creating images, so it doesn't hurt to include that functionality and there's no purpose for a separate singularity-runtime package anymore.

File: dist/rpm/singularity.spec.in
Comment (DrDaveD) [reply dist/rpm/singularity.spec.in]: The reason for introducing the dual package is that RHEL 7.6 is going to have full support for unprivileged mount namespaces, so having a separate unprivileged package makes sense.

File: mlocal/frags/Makefile.stub
Comment (jmstover) [reply mlocal/frags/Makefile.stub:19]: Yes. This was more of a general comment on the wanted makeit usage, nothing against how it was implemented there.

File: dist/rpm/singularity.spec.in
Comment (jmstover) [reply dist/rpm/singularity.spec.in]: > My preference is to make it be as close to the upstream as possible

That's fine. But there are things that you _want_ that don't fit in with what we want upstream. We don't _want_ to be in packaging. We want just a simple example that is usable, if wanted, but doesn't try to be the end all position on how it should be. It's an _example_ that can be used to flesh out a more site specific version if wanted by that admin.

> so having a separate unprivileged package makes sense

If you're doing that, I think you should have reversed it. So instead you have a `package priv` (i.e. singularity-priv) and that includes only the `starter-suid` binary. Everything else is in the main singularity package. How you have it, when you do:

As it is in the PR, if you do `yum install singularity` ... you're installing the *SUID* component. You'd need `yum install singularity-unpriv` for the main Singularity binary. Not what you would expect to happen.

And again, from our point of view. We want out of the packaging market. We want a bare bones and simple example file. It builds an RPM. Nothing fancy, but is a base upon which others can expand on if they wish.

Comment (DrDaveD) [reply dist/rpm/singularity.spec.in]: Could the singularity core team look at it this way: let @drdaved maintain the example spec file in the distribution?  You don't have to worry about it; let any questions & reviews go to me.  EPEL is by far the most widely used distribution of rpm so I think it makes sense to make it be the base-level example.  Site admins can modify that as they like then.  My concern with not doing that and including it in travis builds is that packaging frequently gets broken during the development process.  I experienced that many times with 2.x before singularity travis started building rpms.

With the PR as it is, if someone does `yum install singularity` they automatically get both packages, because the singularity package requires singularity-unpriv of the same version number.  I don't think it's a good idea to put the suid binary only in a singularity-priv package because then some types of basic functionality will not work with just the singularity package.  I think the name "singularity" should contain the full suite, or there will be lots of support questions from people who are just starting out.  Let the people who know that they can live without the suid binary go through the extra effort to learn to use a different package name.

Comment (GodloveD) [reply dist/rpm/singularity.spec.in]: >Could the singularity core team look at it this way: let @DrDaveD maintain the example spec file in the distribution? You don't have to worry about it; let any questions & reviews go to me.

I think that is a great idea, but I think the spec file(s) belong(s) in another repo where @DrDaveD has the ability to merge.  

Only Sylabs-ers have merge permissions in this repo and that is not going to change anytime soon.  So that means Sylabs-ers must review anything that goes into this repo.  We have a policy of 2 reviews per merge.  That is why we don't want a complicated set of spec files living here and frequent packaging PRs coming in.  It's just too big an investment of time.  

@DrDaveD why not create another repo where these and other spec files could live?  And Singularity docs could point folks there?

Comment (jmstover) [reply dist/rpm/singularity.spec.in]: @DrDaveD ... if you're willing to remove the `package unpriv` here in this PR, I'm willing to approve it (and wait for @yhcote to speak up on the makeit parts).

From my point of view, overall it *is* a good PR. It's just that splitting of the RPM package that we're really not wanting to have in the upstream version that's causing the biggest push back from me.

Comment (DrDaveD) [reply dist/rpm/singularity.spec.in]: It's already in a git repo at EPEL (not github currently), but having it in a different repository defeats the purpose of using travis to prevent development from breaking the packaging.  I'm sure you wouldn't want your travis to read from a different repository, would you?

I know that the singularityware/singularity github repo requires two singularity team members to say they approve but that could just be a rubber-stamp kind of review, it doesn't have to take much of their time.

As I said, if I must I will remove the extra "unpriv" package.  Just tell me if I must.

Comment (DrDaveD) [reply dist/rpm/singularity.spec.in]: Ok I removed the unpriv packaging.

**Review by jmstover — APPROVED** [review]

File: dist/rpm/singularity.spec.in
Comment (jmstover) [reply dist/rpm/singularity.spec.in]: Thank you very much.

File: dist/rpm/singularity.spec.in
Comment (jmstover) [inline dist/rpm/singularity.spec.in]: Should GOROOT be set here to:  `GOROOT=${PWD}/go`  ?

Comment (DrDaveD) [reply dist/rpm/singularity.spec.in]: I have never had to set GOROOT.  It must default to `` `dirname $0` `` in the go binary.

On the other hand I guess if the user has set it then that would override the default so you're probably right.  I will add it.

Comment (DrDaveD) [reply dist/rpm/singularity.spec.in]: ok I made that change, and rebased again on the master branch.

**Comment by DrDaveD** [comment]
@yhcote This is waiting on you.  I don't have permission to officially request you as a reviewer.

**Review by yhcote — COMMENTED** [review]
Hi @DrDaveD, I looked at your PR and I think it's pretty much good to go, I took the liberty of creating a commit on top of your PR in my fork branch called "pr1932":
https://github.com/yhcote/singularity/tree/pr1932

Could you please merge that commit in your PR branch so that we can merge #1932?

Cheers.

**Comment by DrDaveD** [comment]
@yhcote I don't want to move the setting of the default package_version because then when a build is done outside of a git clone area (as for example the rpm spec file does) it causes git to print an error message.  I will include the rest of the changes from your commit.

**Comment by DrDaveD** [comment]
Ok I included most of @yhcote's suggested change and rebased again against the current master.

**Comment by yhcote** [comment]
Hi @DrDaveD,

I updated https://github.com/yhcote/singularity/tree/pr1932 to palliate to your concerns.
I think you can merge that code safely in yours here and we'll merge it!

Cheers!

**Comment by DrDaveD** [comment]
Ok @yhcote I included your requested change (and rebased on master again).

File: makeit/mconfig
Comment (yhcote) [inline makeit/mconfig]: All vars above including package_version don't need to be exported with how the build system embeds other units. We can safely remove lines 24 and 25. Cheers.

**Review by yhcote — CHANGES_REQUESTED** [review]

File: makeit/mconfig
Comment (DrDaveD) [reply makeit/mconfig]: Ok I removed the export.

**Review by yhcote — APPROVED** [review]
Looks good to me. Thanks @DrDaveD !

**Pull request merged**
